CORDA-3197 Fix flow has been waiting message

kyriathar opened the pull request:
FlowMonitor would consider as starting time of a "waiting" flow  (suspended flow), the time of its initial creation. That way, it would for e.g. log as "waiting" a flow that could be actually running at that moment and its execution time exceeded the FlowMonitor#suspensionLoggingThreshold.

What we would want instead is the flow to somehow update its current execution time, and therefore FlowMonitor to actually start counting always from that updated time. 

To achieve this, we added a timestamp property to Checkpoint getting a new Instant.now() value at every Checkpoint instantiation/ instantiation by copy. FlowMonitor is now using this new property (Checkpoint#timestamp) as the starting point of a potential suspension. 

That alone though would not be enough, because we could for e.g. have a checkpoint before some long taking task and that could be mistakenly considered by FlowMonitor a suspension period while its not, the flow is still running. 

To overcome the above, FlowMonitor additionally, uses StateMachineState#isFlowResumed to determine if a flow is actually suspended, e.g. waiting for counter parties messages. It leaves out flows that are actually running.

kyriathar-r3 committed:
Added a timestamp property to Checkpoint getting a new Instant.now() value at every Checkpoint instantiation/ copy instantiation. FlowMonitor is now using this new property (Checkpoint#timestamp) and StateMachineState#isFlowResumed to determine which flows are actually suspended. It leaves out flows that are doing work in their FlowLogic#call method.

mnesbit commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineState.kt:
```diff
@@ -62,6 +63,9 @@ data class Checkpoint(
         val numberOfSuspends: Int
 ) {
+
+    val timestamp: Instant = Instant.now()
```
We could make this a var rather than the copy we use above

lankydan replied:
I think the current way works quite nicely, meaning we never need to worry about setting the timestamp as it is done automatically when we change the checkpoint for other reasons.

mnesbit commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -727,7 +727,10 @@ class SingleThreadedStateMachineManager(
+                        // Do a trivial checkpoint copy below, to update the Checkpoint#timestamp value. This is for FlowMonitor.
+                        // We need to refresh the suspension starting time of a to be re-executed from checkpoint flow.
+                        // This applies in case of an e.g. Node startUp after a long period, and for 'Started' flows only (Please check FlowMonitor's flow logging criteria).
+                        checkpoint = checkpoint.copy(),
```
When we merge to ent we also have to remember to include the equivalent change on MultiThreadedStateMachine

blsemo commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineState.kt:
```diff
@@ -62,6 +63,9 @@ data class Checkpoint(
         val errorState: ErrorState,
         val numberOfSuspends: Int
 ) {
+
```
Rather than having a non-constructor val in a data class (they can be tricky...) and relying on a copy function where it isn't entirely obvious how shallow or deep the copy is (we obviously don't want a deep, copy), I would suggest something like this:
```
/**
 * @param invocationContext the initiator of the flow.
 * @param ourIdentity the identity the flow is run as.
 * @param sessions map of source session ID to session state.
 * @param subFlowStack the stack of currently executing subflows.
 * @param flowState the state of the flow itself, including the frozen fiber/FlowLogic.
 * @param errorState the "dirtiness" state including the involved errors and their propagation status.
 * @param numberOfSuspends the number of flow suspends due to IO API calls.
 * @param timeStamp timeStamp of last suspension
 */
data class Checkpoint(
        val invocationContext: InvocationContext,
        val ourIdentity: Party,
        val sessions: SessionMap, // This must preserve the insertion order!
        val subFlowStack: List<SubFlow>,
        val flowState: FlowState,
        val errorState: ErrorState,
        val numberOfSuspends: Int,
        val timeStamp: Instant
) {

    companion object {

        fun create(
                invocationContext: InvocationContext,
                flowStart: FlowStart,
                flowLogicClass: Class<FlowLogic<*>>,
                frozenFlowLogic: SerializedBytes<FlowLogic<*>>,
                ourIdentity: Party,
                subFlowVersion: SubFlowVersion,
                isEnabledTimedFlow: Boolean
        ): Try<Checkpoint> {
            return SubFlow.create(flowLogicClass, subFlowVersion, isEnabledTimedFlow).map { topLevelSubFlow ->
                Checkpoint(
                        invocationContext = invocationContext,
                        ourIdentity = ourIdentity,
                        sessions = emptyMap(),
                        subFlowStack = listOf(topLevelSubFlow),
                        flowState = FlowState.Unstarted(flowStart, frozenFlowLogic),
                        errorState = ErrorState.Clean,
                        numberOfSuspends = 0,
                        timeStamp = Instant.now()
                )
            }
        }
    }
}

/** 
 * Taking a shallow copy of a checkpoint with an updated timestamp.
 */
fun Checkpoint.withUpdatedTimeStamp(): Checkpoint{
    return Checkpoint(
            this.invocationContext,
            this.ourIdentity,
            this.sessions,
            this.subFlowStack,
            this.flowState,
            this.errorState,
            this.numberOfSuspends,
            Instant.now())
}
```

lankydan replied:
Do we still want to add this `withUpdatedTimeStamp` function?

The nice thing about how it is currently implemented is that the `timestamp` is updated automatically in every place that we `copy` the checkpoint.

If we make this change, then we need to update every place that we call `copy` on a `Checkpoint` to also call this function.

The bare minimum is a comment on the `timestamp` to confirm that calling `copy` will update the timestamp so we remember that this is why it works.

blsemo reviewed (commented):
It looks good in principle - see the comment about the Checkpoint changes, this could be cleaner.

kyriathar-r3 committed:
Cleaner comment

kyriathar-r3 committed:
Broke FlowMonitor#logFlowsWaitingForParty into logFlowsWaitingForParty and waitingFlowsToDurations. This way waitingFlowsToDurations is modular and can be tested.

Made FlowMonitor constructor get StateMachineManager instead of the retrieveFlows lamda. This way FlowMonitor is more consistent as a service, and entire flow filtering process is now being done in FlowMonitor#waitingFlowsToDurations.

Removed "smm as? StateMachineManagerInternal" in AbstractNode#start as it made no sense.

Updated CheckpointDumper to mention the Checkpoint#timestamp when writing the checkpoint as json.

kyriathar-r3 committed:
Added tests for FlowMonitor service.

kyriathar-r3 committed:
Remove old comment

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt:
```diff
@@ -53,14 +53,18 @@ internal class FlowMonitor constructor(private val retrieveFlows: () -> Set<Flow
+                if (suspensionDuration >= suspensionLoggingThreshold)
+                    flow.ioRequest()?.let {
+                        request -> warningMessageForFlowWaitingOnIo(request, flow, suspensionDuration)
+                    }?.let(logger::info)
```
I think we should move the `?.let(logger::info)` into the other `let` block to make it:
```kotlin
flow.ioRequest()?.let { logger.info(warningMessageForFlowWaitingOnIo(it, flow, suspensionDuration)) }
```

kyriathar replied:
valid point, FlowMonitor#warningMessageForFlowWaitingOnIo cannot return null.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt:
```diff
@@ -82,9 +86,12 @@ internal class FlowMonitor constructor(private val retrieveFlows: () -> Set<Flow
-    private fun FlowStateMachineImpl<*>.ongoingDuration(now: Instant) = Duration.between(createdAt(), now)
+    private fun FlowStateMachineImpl<*>.ongoingDuration(now: Instant): Duration {
+        val lastCheckpointTimestamp = transientState?.value?.checkpoint?.timestamp ?: return Duration.ZERO
+        return Duration.between(lastCheckpointTimestamp, now)
```
Would it look better as?:

```kotlin
private fun FlowStateMachineImpl<*>.ongoingDuration(now: Instant): Duration {
    return transientState?.value?.checkpoint?.timestamp?.let { Duration.between(it, now) } ?: return Duration.ZERO
}
```

Not sure it really makes a difference, this is just a small thing.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt:
```diff
@@ -12,7 +14,7 @@ import java.util.concurrent.Executors
 import java.util.concurrent.TimeUnit
 
-internal class FlowMonitor constructor(private val retrieveFlows: () -> Set<FlowStateMachineImpl<*>>,
+internal class FlowMonitor constructor(private val smm: StateMachineManager,
```
While we are changing this constructor, can we correct the indentation 😃

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt:
```diff
@@ -49,18 +51,25 @@ internal class FlowMonitor constructor(private val retrieveFlows: () -> Set<Flow
-    private fun warningMessageForFlowWaitingOnIo(request: FlowIORequest<*>, flow: FlowStateMachineImpl<*>, now: Instant): String {
-        val message = StringBuilder("Flow with id ${flow.id.uuid} has been waiting for ${flow.ongoingDuration(now).toMillis() / 1000} seconds ")
+    @VisibleForTesting
+    fun waitingFlowsToDurations(suspensionLoggingThreshold: Duration): Set<Pair<FlowStateMachineImpl<*>, Duration>> {
```
I think we could return the `Sequence` from this function to save the extra iteration in `logFlowsWaitingForParty` (as we only split this function out to make testing easier).

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt:
```diff
@@ -49,18 +51,25 @@ internal class FlowMonitor constructor(private val retrieveFlows: () -> Set<Flow
-    private fun warningMessageForFlowWaitingOnIo(request: FlowIORequest<*>, flow: FlowStateMachineImpl<*>, now: Instant): String {
-        val message = StringBuilder("Flow with id ${flow.id.uuid} has been waiting for ${flow.ongoingDuration(now).toMillis() / 1000} seconds ")
+    @VisibleForTesting
+    fun waitingFlowsToDurations(suspensionLoggingThreshold: Duration): Set<Pair<FlowStateMachineImpl<*>, Duration>> {
```
Would `getWaitingFlowDurations` be a better name?

kyriathar replied:
Since we use this method like so: for (var in method), perhaps something like "waitingFlowDurations"?

lankydan commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -500,22 +500,17 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
-                flowMonitor.start()
-            }
-
+            val flowMonitor = FlowMonitor(smm,
```
Good that check made no sense.

lankydan commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -500,22 +500,17 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
-
+            val flowMonitor = FlowMonitor(smm,
+                    configuration.flowMonitorPeriodMillis,
+                    configuration.flowMonitorSuspensionLoggingThresholdMillis)
```
Small formatting thing. Could you put `smm` on a new line and the last `)` on a new line as well.

lankydan commented on node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt:
```diff
@@ -184,7 +184,7 @@ class CheckpointDumper(private val checkpointStorage: CheckpointStorage, private
                 flowCallStack = flowCallStack,
                 suspendedOn = (flowState as? FlowState.Started)?.flowIORequest?.toSuspendedOn(
-                        suspendedTimestamp(),
+                        timestamp,
```
Did you manage to test that this worked? A manual test should be fine for this.

kyriathar replied:
This one was tested manually, it seems to be working okay.

kyriathar-r3 committed:
1. FLowMonitor#waitingFlowDurations now returns a Sequence to have an iteration less.
It used to be, one iteration from returning a Set from FLowMonitor#waitingFlowDurations plus one iteration from FlowMonitor#logFlowsWaitingForParty.

2. Code reformattings

kyriathar-r3 committed:
1. Remove constructor keyword from FlowMonitor

2. Code reformattings

3. Update detekt baseline

mnesbit commented:
What is holding up this PR?

lankydan reviewed (dismissed):
LGTM, all changes made.

kyriathar-r3 committed:
Resolve conflict in Detekt baseline
